langchain/retrievers/pinecone_self_query.py

```diff
@@ -0,0 +1,133 @@
+import json
+from typing import Any, Dict, List, cast
+
+from pydantic import BaseModel, Field, root_validator
+
+from langchain import LLMChain, PromptTemplate
+from langchain.agents.agent_toolkits import VectorStoreInfo
+from langchain.llms import BaseLLM
+from langchain.retrievers.pinecone_self_query_prompt import (
+    PineconeSelfQueryOutputParser,
+    pinecone_example,
+    pinecone_format_instructions,
+    self_query_prompt,
+)
+from langchain.schema import BaseRetriever, Document
+from langchain.vectorstores import Pinecone
+
+
+class MetadataFieldInfo(BaseModel):
+    """Information about a vectorstore metadata field."""
+
+    name: str
+    description: str
+    examples: List
+    type: str
+
+    class Config:
+        """Configuration for this pydantic object."""
+
+        arbitrary_types_allowed = True
+
+
+class VectorStoreExtendedInfo(VectorStoreInfo):
+    """Extension of VectorStoreInfo that includes info about metadata fields."""
+
+    metadata_field_info: List[MetadataFieldInfo]
+    """Map of metadata field name to info about that field."""
+
+
+def _format_metadata_field_info(info: List[MetadataFieldInfo]) -> str:
+    info_dicts = {}
+    for i in info:
+        i_dict = dict(i)
+        info_dicts[i_dict.pop("name")] = i_dict
+    return json.dumps(info_dicts, indent=2).replace("{", "{{").replace("}", "}}")
+
+
+class PineconeSelfQueryRetriever(BaseRetriever, BaseModel):
```

Comment:
technically nothing here except the default prompts here is Pinecone-specific. on the one hand limiting scope is nice, on the other it may make backwards compatibility annoying if we do create a generic version of this sometime soon and have to maintain this one specific version. thoughts @hwchase17

Reply:
at a glance seems easier to abstract later rather than guess at a base class given that it presumably could evolve into much more general types of retrievers